README.md
# Consistent Overhead Byte Stuffing (COBS) implementation in Dart.

This package is a Dart port of original COBS implementation in C
by Craig McQueen and influenced by the Java implementation by
TheMarpe (https://github.com/themarpe)

# Intro

A library is provided, which contains functions for encoding and decoding
according to COBS methods.


## What Is COBS?

COBS is a method of encoding a packet of bytes into a form that contains no
bytes with value zero (0x00). The input packet of bytes can contain bytes
in the full range of 0x00 to 0xFF. The COBS encoded packet is guaranteed to
generate packets with bytes only in the range 0x01 to 0xFF. Thus, in a
communication protocol, packet boundaries can be reliably delimited with 0x00
bytes.

The COBS encoding does have to increase the packet size to achieve this
encoding. However, compared to other byte-stuffing methods, the packet size
increase is reasonable and predictable. COBS always adds 1 byte to the
message length. Additionally, for longer packets of length *n*, it *may* add
n/254 (rounded down) additional bytes to the encoded packet size.

For example, compare to the PPP protocol, which uses 0x7E bytes to delimit
PPP packets. The PPP protocol uses an "escape" style of byte stuffing,
replacing all occurrences of 0x7E bytes in the packet with 0x7D 0x5E. But that
byte-stuffing method can potentially double the size of the packet in the
worst case. COBS uses a different method for byte-stuffing, which has a much
more reasonable worst-case overhead.

For more details about COBS, see the references [1][1] [2][2].


## References

[Consistent Overhead Byte Stuffing][1]\
Stuart Cheshire and Mary Baker\
IEEE/ACM Transations on Networking, Vol. 7, No. 2, April 1999

[PPP Consistent Overhead Byte Stuffing (COBS)][2]\
PPP Working Group Internet Draft<br>James Carlson, IronBridge Networks\
Stuart Cheshire and Mary Baker, Stanford University, November 1997

[1]: http://www.stuartcheshire.org/papers/COBSforToN.pdf
[2]: http://tools.ietf.org/html/draft-ietf-pppext-cobs-00

CHANGELOG.md
## [0.1.0] - 28 April 2019

* Implement Stream decoding
* Add `withZero` optional parameter to `encodeDstBufMaxLen` and `encodeCOBS()` 
  to append a zero byte to the end of the encoded packet.
* Add dartdoc comments

## [0.0.1] - 28 April 2019

* Initial release of Dart implementation of COBS algorithm.


test/cobs_test.dart
import 'dart:async';
import 'dart:typed_data';

import 'package:collection/collection.dart';
import 'package:test/test.dart';

import 'package:cobs/cobs.dart';

void main() {

  Function eq = const ListEquality().equals;

  group('Encode and decode byte data using COBS', () {
    test('encodeCOBS returns error with null source data', () {
      ByteData source;
      var encoded = ByteData(1);
      EncodeResult encodeResult = encodeCOBS(encoded, source);
      expect(encodeResult.status, EncodeStatus.NULL_POINTER);
      expect(encodeResult.outLen, 0);
    });

    test('decodeCOBS returns error with null source data', () {
      ByteData source;
      var decoded = ByteData(1);
      DecodeResult decodeResult = decodeCOBS(decoded, source);
      expect(decodeResult.status, DecodeStatus.NULL_POINTER);
      expect(decodeResult.outLen, 0);
    });

    test('encode / decode all zeroes data', () {
      var data = new Uint8List.fromList([ 0, 0, 0, 0 ]);
      var initData = ByteData.view(data.buffer);
      ByteData encoded = ByteData(encodeDstBufMaxLen(initData.lengthInBytes));
      expect(encoded.lengthInBytes, 5);

      EncodeResult encodeResult = encodeCOBS(encoded, initData);
      expect(encodeResult.outLen, 5);
      expect(encodeResult.status, EncodeStatus.OK);

      ByteData decoded = ByteData(decodeDstBufMaxLen(encoded.lengthInBytes));
      expect(decoded.lengthInBytes, 4);

      DecodeResult decodeResult = decodeCOBS(decoded, encoded);
      expect(decodeResult.outLen, 4);
      expect(decodeResult.status, DecodeStatus.OK);
      expect(
          eq(initData.buffer.asUint8List(), decoded.buffer.asUint8List()),
          true);
    });

    test('encode / decode data with no zeroes', () {
      var data = new Uint8List.fromList([ 1, 2, 3, 4, 5 ]);
      var initData = ByteData.view(data.buffer);
      ByteData encoded = ByteData(encodeDstBufMaxLen(initData.lengthInBytes));
      expect(encoded.lengthInBytes, 6);

      EncodeResult encodeResult = encodeCOBS(encoded, initData);
      expect(encodeResult.outLen, 6);
      expect(encodeResult.status, EncodeStatus.OK);

      ByteData decoded = ByteData(decodeDstBufMaxLen(encoded.lengthInBytes));
      expect(decoded.lengthInBytes, 5);

      DecodeResult decodeResult = decodeCOBS(decoded, encoded);
      expect(decodeResult.outLen, 5);
      expect(decodeResult.status, DecodeStatus.OK);
      expect(
          eq(initData.buffer.asUint8List(), decoded.buffer.asUint8List()),
          true);
    });

    test('encode / decode data with zero at end', () {
      var data = new Uint8List.fromList([ 1, 2, 3, 4, 5, 0 ]);
      var initData = ByteData.view(data.buffer);
      ByteData encoded = ByteData(encodeDstBufMaxLen(initData.lengthInBytes));
      expect(encoded.lengthInBytes, 7);

      EncodeResult encodeResult = encodeCOBS(encoded, initData);
      expect(encodeResult.outLen, 7);
      expect(encodeResult.status, EncodeStatus.OK);

      ByteData decoded = ByteData(decodeDstBufMaxLen(encoded.lengthInBytes));
      expect(decoded.lengthInBytes, 6);

      DecodeResult decodeResult = decodeCOBS(decoded, encoded);
      expect(decodeResult.outLen, 6);
      expect(decodeResult.status, DecodeStatus.OK);
      expect(
          eq(initData.buffer.asUint8List(), decoded.buffer.asUint8List()),
          true);
    });

    test('encode / decode data with embedded zero', () {
      var data = new Uint8List.fromList([ 1, 2, 3, 4, 5, 0, 6, 7, 8, 9 ]);
      var initData = ByteData.view(data.buffer);
      ByteData encoded = ByteData(encodeDstBufMaxLen(initData.lengthInBytes));
      expect(encoded.lengthInBytes, 11);

      EncodeResult encodeResult = encodeCOBS(encoded, initData);
      expect(encodeResult.outLen, 11);
      expect(encodeResult.status, EncodeStatus.OK);

      ByteData decoded = ByteData(decodeDstBufMaxLen(encoded.lengthInBytes));
      expect(decoded.lengthInBytes, 10);

      DecodeResult decodeResult = decodeCOBS(decoded, encoded);
      expect(decodeResult.outLen, 10);
      expect(decodeResult.status, DecodeStatus.OK);
      expect(
          eq(initData.buffer.asUint8List(), decoded.buffer.asUint8List()),
          true);
    });

    test('encode / decode data with embedded zero and zero at end', () {
      var data = new Uint8List.fromList([ 1, 2, 3, 4, 5, 0, 6, 7, 8, 9, 0 ]);
      var initData = ByteData.view(data.buffer);
      ByteData encoded = ByteData(encodeDstBufMaxLen(initData.lengthInBytes));
      expect(encoded.lengthInBytes, 12);

      EncodeResult encodeResult = encodeCOBS(encoded, initData);
      expect(encodeResult.outLen, 12);
      expect(encodeResult.status, EncodeStatus.OK);

      ByteData decoded = ByteData(decodeDstBufMaxLen(encoded.lengthInBytes));
      expect(decoded.lengthInBytes, 11);

      DecodeResult decodeResult = decodeCOBS(decoded, encoded);
      expect(decodeResult.outLen, 11);
      expect(decodeResult.status, DecodeStatus.OK);
      expect(
          eq(initData.buffer.asUint8List(), decoded.buffer.asUint8List()),
          true);
    });

    test('encode / decode data with 253 bytes', () {
      var data = new Uint8List.fromList([ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253 ]);
      var initData = ByteData.view(data.buffer);
      ByteData encoded = ByteData(encodeDstBufMaxLen(initData.lengthInBytes));
      expect(encoded.lengthInBytes, 254);

      EncodeResult encodeResult = encodeCOBS(encoded, initData);
      expect(encodeResult.outLen, 254);
      expect(encodeResult.status, EncodeStatus.OK);

      ByteData decoded = ByteData(decodeDstBufMaxLen(encoded.lengthInBytes));
      expect(decoded.lengthInBytes, 253);

      DecodeResult decodeResult = decodeCOBS(decoded, encoded);
      expect(decodeResult.outLen, 253);
      expect(decodeResult.status, DecodeStatus.OK);
      expect(
          eq(initData.buffer.asUint8List(), decoded.buffer.asUint8List()),
          true);
    });

    test('encode / decode data with 254 bytes', () {
      var data = new Uint8List.fromList([ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254 ]);
      var initData = ByteData.view(data.buffer);
      ByteData encoded = ByteData(encodeDstBufMaxLen(initData.lengthInBytes));
      expect(encoded.lengthInBytes, 255);

      EncodeResult encodeResult = encodeCOBS(encoded, initData);
      expect(encodeResult.outLen, 255);
      expect(encodeResult.status, EncodeStatus.OK);

      ByteData decoded = ByteData(decodeDstBufMaxLen(encoded.lengthInBytes));
      expect(decoded.lengthInBytes, 254);

      DecodeResult decodeResult = decodeCOBS(decoded, encoded);
      expect(decodeResult.outLen, 254);
      expect(decodeResult.status, DecodeStatus.OK);
      expect(
          eq(initData.buffer.asUint8List(), decoded.buffer.asUint8List()),
          true);
    });

    test('encode / decode data with 255 bytes', () {
      var data = new Uint8List.fromList([ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255 ]);
      var initData = ByteData.view(data.buffer);
      ByteData encoded = ByteData(encodeDstBufMaxLen(initData.lengthInBytes));
      expect(encoded.lengthInBytes, 257);

      EncodeResult encodeResult = encodeCOBS(encoded, initData);
      expect(encodeResult.outLen, 257);
      expect(encodeResult.status, EncodeStatus.OK);

      ByteData decoded = ByteData(decodeDstBufMaxLen(encoded.lengthInBytes));
      expect(decoded.lengthInBytes, 256);

      DecodeResult decodeResult = decodeCOBS(decoded, encoded);
      expect(decodeResult.outLen, 255);
      expect(decodeResult.status, DecodeStatus.OK);
      expect(
          eq(initData.buffer.asUint8List(), decoded.buffer.asUint8List(0, 255)),
          true);
    });

    test('encode withZero appends zero byte', () {
      var data = new Uint8List.fromList([ 1, 2, 3, 4, 5]);
      var initData = ByteData.view(data.buffer);
      ByteData encoded = ByteData(
          encodeDstBufMaxLen(initData.lengthInBytes, withZero: true));
      expect(encoded.lengthInBytes, 7);

      EncodeResult encodeResult = encodeCOBS(encoded, initData, withZero: true);
      expect(encodeResult.outLen, 7);
      expect(encodeResult.status, EncodeStatus.OK);
      expect(encoded.getUint8(6), 0x00);
    });

  });

  group('Decode ByteData stream using decodeCOBSStream', () {
    test('decode stream of a complete packet', () async {
      var data = new Uint8List.fromList([ 1, 2, 3, 4, 5, 0, 6, 7, 8, 9, 0 ]);
      var initData = ByteData.view(data.buffer);
      ByteData encoded = ByteData(encodeDstBufMaxLen(initData.lengthInBytes, withZero: true));
      expect(encoded.lengthInBytes, 13);

      var controller = new StreamController<ByteData>();
      var decoder = decodeCOBSStream(controller.stream);

      EncodeResult encodeResult = encodeCOBS(encoded, initData, withZero: true);
      expect(encodeResult.outLen, 13);
      expect(encodeResult.status, EncodeStatus.OK);

      controller.sink.add(encoded);

      await decoder.listen(expectAsync1((decoded) {
        expect(
            eq(decoded.buffer.asUint8List(), initData.buffer.asUint8List()),
            true);
      }, count: 1));
    });

    test('decode stream of two incomplete packets', () async {
      var data = new Uint8List.fromList([ 1, 2, 3, 4, 5, 0, 6, 7, 8, 9, 0 ]);
      var initData = ByteData.view(data.buffer);

      ByteData encoded = ByteData(encodeDstBufMaxLen(initData.lengthInBytes, withZero: true));
      expect(encoded.lengthInBytes, 13);

      EncodeResult encodeResult = encodeCOBS(encoded, initData, withZero: true);
      expect(encodeResult.outLen, 13);
      expect(encodeResult.status, EncodeStatus.OK);

      var first = encoded.buffer.asByteData(0, 4);
      var second = encoded.buffer.asByteData(4);

      var controller = new StreamController<ByteData>();
      var decoder = decodeCOBSStream(controller.stream);

      controller.sink.add(first);
      controller.sink.add(second);

      await decoder.listen(expectAsync1((decoded) {
        expect(
            eq(decoded.buffer.asUint8List(), initData.buffer.asUint8List()),
            true);
      }, count: 1));
    });

  });
}


lib/cobs.dart
library cobs;

import 'dart:typed_data';

/// [encodeCOBS()] status types
enum EncodeStatus {
  OK,
  NULL_POINTER,
  OUT_BUFFER_OVERFLOW,
}

/// Status and length container of [encodeCOBS()] result.
class EncodeResult {
  int outLen;
  EncodeStatus status;
}

/// [decodeCOBS()] status types.
enum DecodeStatus {
  OK,
  NULL_POINTER,
  OUT_BUFFER_OVERFLOW,
  ZERO_BYTE_IN_INPUT,
  INPUT_TOO_SHORT,
}

/// Status and length container of [decodeCOBS()] result.
class DecodeResult {
  int outLen;
  DecodeStatus status;
}

/// Determine maximum encoded byte length for source containing [srcLen] bytes.
///
/// If [withZero] is true, the length is increased by 1 to allow room for a 0x00
/// byte to be appended to the end
int encodeDstBufMaxLen(int srcLen, {bool withZero=false}) {
  return srcLen + (srcLen + 253)~/254 + (withZero ? 1 : 0);
}

/// Determine maximum decoded byte length for source containing [srcLen] bytes.
int decodeDstBufMaxLen(int srcLen){
  return srcLen == 0 ? 0 : srcLen - 1;
}

/// Encode [source] to [encoded] using COBS and return [EncodeResult] status.
///
/// If [withZero] is true, a 0x00 byte is appended to [encoded]
/// The [EncodeResult] instance returned will include the actual length of the
/// encoded byte array in [outLen] and the status of the encoding attempt in
/// [status].
EncodeResult encodeCOBS(ByteData encoded, ByteData source,
    {bool withZero=false}) {
  EncodeResult result = new EncodeResult();
  result.outLen = 0;
  result.status = EncodeStatus.OK;

  if (encoded == null || source == null) {
    result.status = EncodeStatus.NULL_POINTER;
    return result;
  }

  int encodedWriteCounter = 1;
  int encodedCodeWriteCounter = 0;
  int encodedEndCounter = encoded.lengthInBytes;

  int sourceCounter = 0;
  int sourceEndCounter = source.lengthInBytes;

  int searchLen = 1;

  if (source.lengthInBytes != 0) {
    /* Iterate over the source bytes */
    while (true) {
      /* Check for running out of output buffer space */
      if (encodedWriteCounter >= encodedEndCounter) {
        result.status = EncodeStatus.OUT_BUFFER_OVERFLOW;
        break;
      }

      int sourceByte = source.getUint8(sourceCounter++);
      if (sourceByte == 0) {
        /* We found a zero byte */
        encoded.setUint8(encodedCodeWriteCounter, searchLen & 0xFF);
        encodedCodeWriteCounter = encodedWriteCounter++;
        searchLen = 1;
        if (sourceCounter >= sourceEndCounter) {
          break;
        }
      } else {
        /* Copy the non-zero byte to the destination buffer */
        encoded.setUint8(encodedWriteCounter++, sourceByte & 0xFF);
        searchLen++;

        if (sourceCounter >= sourceEndCounter) {
          break;
        }

        if (searchLen == 0xFF) {
          /* We have a long string of non-zero bytes, so we need
           * to write out a length code of 0xFF. */
          encoded.setUint8(encodedCodeWriteCounter, searchLen & 0xFF);
          encodedCodeWriteCounter = encodedWriteCounter++;
          searchLen = 1;
        }
      }
    }
  }

  if (withZero) {
    encoded.setUint8(encodedWriteCounter++, 0x00);
  }

  /* We've reached the end of the source data (or possibly run out of output buffer)
   * Finalise the remaining output. In particular, write the code (length) byte.
   * Update the pointer to calculate the final output length.
   */
  if (encodedCodeWriteCounter >= encodedEndCounter) {
    /* We've run out of output buffer to write the code byte. */
    result.status = EncodeStatus.OUT_BUFFER_OVERFLOW;
    encodedWriteCounter = encodedEndCounter;
  } else {
    /* Write the last code (length) byte. */
    encoded.setUint8(encodedCodeWriteCounter, searchLen & 0xFF);
  }

  /* Calculate the output length, from the value of dst_code_write_ptr */
  result.outLen = encodedWriteCounter;

  return result;
}


/// Decode [source] to [decoded] using COBS and return [DecodeResult] status.
///
/// The [DecodeResult] instance returned will include the actual length of the
/// decoded byte array in [outLen] and the status of the decoding attempt in
/// [status].
DecodeResult decodeCOBS(ByteData decoded, ByteData source){
  DecodeResult result = new DecodeResult();
  result.outLen = 0;
  result.status = DecodeStatus.OK;

  /* First, do a NULL check and return immediately if it fails. */
  if (decoded == null || source == null) {
    result.status = DecodeStatus.NULL_POINTER;
    return result;
  }

  int sourceCounter = 0;
  int sourceEndCounter = source.lengthInBytes;
  int decodedEndCounter = decoded.lengthInBytes;
  int decodedWriteCounter = 0;
  int remainingBytes;
  int sourceByte;
  int i;
  int lengthCode;

  if (source.lengthInBytes != 0) {
    while (true) {
      lengthCode = source.getUint8(sourceCounter++);
      if (lengthCode == 0) {
        result.status = DecodeStatus.ZERO_BYTE_IN_INPUT;
        break;
      }
      lengthCode--;

      /* Check length code against remaining input bytes */
      remainingBytes = sourceEndCounter - sourceCounter;
      if (lengthCode > remainingBytes) {
        result.status = DecodeStatus.INPUT_TOO_SHORT;
        lengthCode = remainingBytes;
      }

      /* Check length code against remaining output buffer space */
      remainingBytes = decodedEndCounter - decodedWriteCounter;
      if (lengthCode > remainingBytes) {
        result.status = DecodeStatus.OUT_BUFFER_OVERFLOW;
        lengthCode = remainingBytes;
      }

      for (i = lengthCode; i != 0; i--) {
        sourceByte = source.getUint8(sourceCounter++);
        if (sourceByte == 0) {
          result.status = DecodeStatus.ZERO_BYTE_IN_INPUT;
        }
        decoded.setUint8(decodedWriteCounter++, sourceByte);
      }

      if (sourceCounter >= sourceEndCounter) {
        break;
      }

      /* Add a zero to the end */
      if (lengthCode != 0xFE) {
        if (decodedWriteCounter >= decodedEndCounter) {
          result.status = DecodeStatus.OUT_BUFFER_OVERFLOW;
          break;
        }
        decoded.setUint8(decodedWriteCounter++, 0);
      }
    }
  }

  result.outLen = decodedWriteCounter;
  return result;
}

/// Decodes a stream of COBS-encoded bytes into packets of decoded bytes.
///
/// The input bytes are provided in chunks through the [source] stream.
Stream<ByteData> decodeCOBSStream(Stream<ByteData> source) async* {
  // Stores any partial packets from the previous chunk.
  var partial = ByteData(0);

  // Wait until a new chunk is available, then process it.
  await for (var chunk in source) {
    // Get index of 0x00 byte if found
    var offset = 0;
    for (var i=0; i<chunk.lengthInBytes; i++) {
      if (chunk.getUint8(i) == 0x00) {
        // Packet delimiter found, now decode it
        int encodedLength = partial.lengthInBytes + i - offset;
        var encoded = ByteData(encodedLength);
        var index = 0;
        for (var ip=0; ip<partial.lengthInBytes; ip++) {
          encoded.setUint8(index, partial.getUint8(ip));
          index++;
        }
        for (var ie=offset; ie<i; ie++) {
          encoded.setUint8(index, chunk.getUint8(ie));
          index++;
        }

        partial = ByteData(0);
        offset = i+1;

        ByteData decoded = ByteData(decodeDstBufMaxLen(encodedLength));
        DecodeResult result = decodeCOBS(decoded, encoded);
        if (result.status == DecodeStatus.OK) {
          yield ByteData.view(decoded.buffer, 0, result.outLen);
        }
      }
    }
    partial = chunk.buffer.asByteData(offset, chunk.lengthInBytes-offset);
  }
}
